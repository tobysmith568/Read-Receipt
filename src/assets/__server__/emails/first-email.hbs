<!DOCTYPE html>
<html>

<head>
  <title>Read Receipt</title>
</head>

<body>
  <p>Hey ðŸ‘‹</p>
  <p>
    You received this email because your email address was entered into the tool
    <a target="_blank" rel="noopener" href="{{domain}}">Read Receipt</a>
    created by <a target="_blank" rel="noopener" href="https://tobysmith.uk">Toby Smith</a>.
  </p>

  <p>
    This email contains a 'tracking pixel'. A tiny invisible image which, when loaded by you opening this email, gave
    away not only that you opened this email, but also some personal information about you and your device.
  </p>

  <p>
    Some email providers are able to block these pixels, but assuming that it was successful, opening this email will
    trigger another email to be sent to you. This second email, which you should have already received by now, contains
    all of the information I was able to gather from you opening this email.
  </p>

  <p>
    This tool stores <b>none</b> of your personal information. It's totally stateless with no cookies,
    databases, logs, or anything else! You can check the source code on
    <a target="_blank" rel="noopener" href="https://github.com/tobysmith568/Read-Receipt">my GitHub</a>.
  </p>

  <p>
    If you have any questions, you can reach out to me at <a
      href="mailto:tobysmith568@hotmail.co.uk">tobysmith568@hotmail.co.uk</a> or find me at <a
      href="https://tobysmith.uk">https://tobysmith.uk</a>.
  </p>

  <p>
    Thanks!
  </p>

  <img src="{{domain}}email/{{urlSafeEmail}}/{{timestamp}}/pixel.png" />
</body>

</html>
