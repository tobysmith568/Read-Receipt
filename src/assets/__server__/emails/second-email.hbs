<!DOCTYPE html>
<html>

<head>
  <title>Read Receipt</title>
</head>

<body>
  <p>Hey ðŸ‘‹</p>
  <p>
    You received this email because your email address was entered into the tool <a href="{{domain}}">Read Receipt</a>
    created by <a href="https://tobysmith.uk">Toby Smith</a> and you opened an email containing a tracking pixel.
  </p>

  <p>
    TODO: Add more text
  </p>

  <p>
    You:
  </p>

  <ul>
    <li>Your Email: {{user.email}}</li>
  </ul>

  <p>
    Times:
  </p>

  <ul>
    <li>First email sent at: {{times.firstEmailTimestamp}}</li>
    <li>This email sent at: {{times.secondEmailTimestamp}}</li>
    <li>Time difference: {{times.timestampDifference}}</li>
  </ul>

  <p>
    Note that this tool stores <b>none</b> of your personal information. It's totally stateless with no cookies,
    databases, logs, or anything else! You can check the source code on
    <a target="_blank" rel="noopener" href="https://github.com/tobysmith568/Read-Receipt">my GitHub</a>.
  </p>

  <p>
    If you have any questions, you can reach out to me at <a
      href="mailto:tobysmith568@hotmail.co.uk">tobysmith568@hotmail.co.uk</a> or find me at <a
      href="https://tobysmith.uk">https://tobysmith.uk</a>.
  </p>

  <p>
    Thanks!
  </p>
</body>

</html>
