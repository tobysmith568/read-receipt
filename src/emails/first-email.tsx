/* eslint-disable jsx-a11y/alt-text */
/* eslint-disable @next/next/no-img-element */
import { FC } from "react";
import { renderToString } from "react-dom/server";

interface Props {
  domain: string;
  urlSafeEmail: string;
  timestamp: number;
}

export const firstEmailAsHtml = (props: Props): string => {
  return `<!DOCTYPE html>
<html>

<head>
  <title>Read Receipt</title>
</head>

<body>
${renderToString(<FirstEmail {...props} />)}
</body>

</html>`;
};

const FirstEmail: FC<Props> = ({ domain, timestamp, urlSafeEmail }) => {
  return (
    <>
      <p>Hey ðŸ‘‹</p>
      <p>
        You received this email because your email address was entered into the tool{" "}
        <a target="_blank" rel="noreferrer noopener" href={domain}>
          Read Receipt
        </a>{" "}
        created by{" "}
        <a target="_blank" rel="noreferrer noopener" href="https://tobysmith.uk">
          Toby Smith
        </a>
        .
      </p>

      <p>
        This email contains a &apos;tracking pixel&apos;; a tiny invisible image. When you opened
        this email, the image loaded and gave away not only that you opened this email, but also
        some personal information about you and your device.
      </p>

      <p>
        Some email providers are able to block these pixels, but assuming that it was successful,
        opening this email will trigger another email to be sent to you. This second email, which
        you should have already received by now, contains all of the information I was able to
        gather from you opening this email.
      </p>

      <p>
        This tool stores <b>none</b> of your personal information. It&apos;s totally stateless with
        no cookies, databases, logs, or anything else! You can check the source code{" "}
        <a
          target="_blank"
          rel="noreferrer noopener"
          href="https://github.com/tobysmith568/Read-Receipt">
          on GitHub
        </a>
        .
      </p>

      <p>
        If you have any questions, you can reach out to me at{" "}
        <a href="mailto:tobysmith568@hotmail.co.uk">tobysmith568@hotmail.co.uk</a> or find me at{" "}
        <a href="https://tobysmith.uk">https://tobysmith.uk</a>.
      </p>

      <p>Thanks!</p>

      <img src={`${domain}/api/open/${urlSafeEmail}/${timestamp}`} />
    </>
  );
};
