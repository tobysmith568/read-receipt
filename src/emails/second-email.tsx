/* eslint-disable jsx-a11y/alt-text */
/* eslint-disable @next/next/no-img-element */
import React, { FC } from "react";
import { renderToString } from "react-dom/server";
import { IpResponse } from "../utils/ip";
import { UserAgentDetails } from "../utils/user-agent";

export interface User {
  email: string;
  ip: string;
}

export interface Times {
  firstEmailTimestamp: string;
  secondEmailTimestamp: string;
  timestampDifference: string;
}

export interface SecondEmailProps {
  domain: string;
  user: User;
  times: Times;
  ipData?: IpResponse;
  userAgentData?: UserAgentDetails;
}

export const secondEmailAsHtml = (props: SecondEmailProps): string => {
  return `<!DOCTYPE html>
<html>

<head>
  <title>Read Receipt</title>
</head>

<body>
${renderToString(<SecondEmail {...props} />)}
</body>

</html>`;
};

const SecondEmail: FC<SecondEmailProps> = ({ domain, user, times, ipData, userAgentData }) => {
  return (
    <>
      <p>Hey ðŸ‘‹</p>
      <p>
        You received this email because your email address was entered into the tool{" "}
        <a href={domain}>Read Receipt</a> created by <a href="https://tobysmith.uk">Toby Smith</a>{" "}
        and you opened an email containing a tracking pixel.
      </p>

      <p>
        Below shows all of the information which my implementation of a tracking pixel is able to
        expose. Note that other people or companies may be able to find out more!
      </p>

      <p>Times:</p>

      <ul>
        <li>First email sent at: {times.firstEmailTimestamp}</li>
        <li>This email sent at: {times.secondEmailTimestamp}</li>
        <li>Time difference: {times.timestampDifference}</li>
      </ul>

      <p>You:</p>
      <ul>
        <li>Email: {user.email}</li>
        <li>IP Address: {user.ip}</li>
      </ul>

      {!!ipData && (
        <>
          <p>
            Your <b>approximate</b> location:
          </p>
          <ul>
            <li>{ipData.city}</li>
            <li>{ipData.zip}</li>
            <li>
              {ipData.regionName} ({ipData.region})
            </li>
            <li>
              {ipData.country} ({ipData.countryCode})
            </li>
            <p>
              <li>
                Lat: {ipData.lat}, Lon: {ipData.lon}
              </li>
            </p>
          </ul>

          <p>Your connection:</p>
          <ul>
            <li>Is Mobile: {ipData.mobile ? "Yes" : "No"}</li>
            <li>Is via Proxy: {ipData.proxy ? "Yes" : "No"}</li>
            <li>Internet provider: {ipData.isp}</li>
          </ul>
        </>
      )}

      {!!userAgentData && (
        <>
          <p>Your device:</p>
          <ul>
            {!!userAgentData.browser && <li>Browser: {userAgentData.browser}</li>}
            {!!userAgentData.version && <li>Version: {userAgentData.version}</li>}
            {!!userAgentData.os && <li>Operating System: {userAgentData.os}</li>}
            {!!userAgentData.platform && <li>Platform: {userAgentData.platform}</li>}
          </ul>
        </>
      )}

      <p>
        Some of this information may be inaccurate, especially the location and device sections.
        This is because some providers attempt to protect their users from tools like the tracking
        pixel hidden in the first email.
      </p>

      <p>
        Reopening the first email you received on different devices or by enabling/disabling WiFi on
        mobile devices will yield different results with different levels of data accuracy. If you
        have multiple emails from different providers you will see different results from each of
        them too, eg, @outlook.com or @gmail.com.
      </p>

      <p>
        This tool stores <b>none</b> of your personal information. It&apos;s totally stateless with
        no cookies, databases, logs, or anything else! You can check the source code{" "}
        <a
          target="_blank"
          rel="noreferrer noopener"
          href="https://github.com/tobysmith568/Read-Receipt">
          on GitHub
        </a>
        .
      </p>

      <p>
        If you have any questions, you can reach out to me at{" "}
        <a href="mailto:tobysmith568@hotmail.co.uk">tobysmith568@hotmail.co.uk</a> or find me at{" "}
        <a href="https://tobysmith.uk">https://tobysmith.uk</a>.
      </p>

      <p>Thanks!</p>
    </>
  );
};
