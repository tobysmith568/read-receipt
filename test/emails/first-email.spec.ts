import { firstEmailAsHtml, FirstEmailProps } from "../../src/emails/first-email";

describe("first email", () => {
  it("should render correctly", () => {
    const props: FirstEmailProps = {
      domain: "https://wherever.this.is.hosted",
      timestamp: 1234567890,
      urlSafeEmail: "this.is.a.url.safe.email.address%40mail.com"
    };

    const result = firstEmailAsHtml(props);

    expect(result).toMatchSnapshot();
  });
});
