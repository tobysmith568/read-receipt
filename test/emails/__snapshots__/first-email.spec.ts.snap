// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`first email should render correctly 1`] = `
"<!DOCTYPE html>
<html>

<head>
  <title>Read Receipt</title>
</head>

<body>
<p>Hey ðŸ‘‹</p><p>You received this email because your email address was entered into the tool<!-- --> <a target=\\"_blank\\" rel=\\"noreferrer noopener\\" href=\\"https://wherever.this.is.hosted\\">Read Receipt</a> <!-- -->created by<!-- --> <a target=\\"_blank\\" rel=\\"noreferrer noopener\\" href=\\"https://tobysmith.uk\\">Toby Smith</a>.</p><p>This email contains a &#x27;tracking pixel&#x27;; a tiny invisible image. When you opened this email, the image loaded and gave away not only that you opened this email, but also some personal information about you and your device.</p><p>Some email providers are able to block these pixels, but assuming that it was successful, opening this email will trigger another email to be sent to you. This second email, which you should have already received by now, contains all of the information I was able to gather from you opening this email.</p><p>This tool stores <b>none</b> of your personal information. It&#x27;s totally stateless with no cookies, databases, logs, or anything else! You can check the source code<!-- --> <a target=\\"_blank\\" rel=\\"noreferrer noopener\\" href=\\"https://github.com/tobysmith568/Read-Receipt\\">on GitHub</a>.</p><p>If you have any questions, you can reach out to me at<!-- --> <a href=\\"mailto:tobysmith568@hotmail.co.uk\\">tobysmith568@hotmail.co.uk</a> or find me at<!-- --> <a href=\\"https://tobysmith.uk\\">https://tobysmith.uk</a>.</p><p>Thanks!</p><img src=\\"https://wherever.this.is.hosted/api/open/this.is.a.url.safe.email.address%40mail.com/1234567890\\"/>
</body>

</html>"
`;
